### PR TITLE
Telemetry revamp part2

### PR DESCRIPTION
**Pull request by thalesmg**
Adds to telemetry:

- MQTT runtime insights
- Advanced MQTT features

### CODE REVIEW

File: apps/emqx_modules/src/emqx_telemetry.erl [hunk 1]
```diff
@@ -401,9 +411,65 @@ vm_specs() ->
...
+            end,
+            {#{}, PrevMetrics0},
+            MetricsToCheck
+        ),
+    {Metrics, State#state{previous_metrics = PrevMetrics}};
+update_mqtt_rates(State) ->
+    {#{}, State}.
+
+advanced_mqtt_features() ->
+    RewriteEnabled = length(emqx:get_config([rewrite], [])) > 0,
```
Comment (zmstone) [inline apps/emqx_modules/src/emqx_telemetry.erl:452]: export an API from emqx_rewrite.erl?
e.g. `emqx_rewrite:is_enabled/0`

File: apps/emqx_modules/src/emqx_telemetry.erl [hunk 2]
```diff
@@ -401,9 +411,65 @@ vm_specs() ->
...
+    AutoSubEnabled = length(emqx:get_config([auto_subscribe, topics], [])) > 0,
```
Comment (zmstone) [inline apps/emqx_modules/src/emqx_telemetry.erl:453]: emqx_auto_subscribe:is_enabled/0

File: apps/emqx_modules/src/emqx_telemetry.erl [hunk 3]
```diff
@@ -401,9 +411,65 @@ vm_specs() ->
...
+    #{
+        retained => bool2int(emqx:get_config([retainer, enable], false)),
```
Comment (zmstone) [inline apps/emqx_modules/src/emqx_telemetry.erl:455]: emqx_retainer:is_enabled.

File: apps/emqx_modules/src/emqx_telemetry.erl [hunk 4]
```diff
@@ -401,9 +411,65 @@ vm_specs() ->
...
+        topic_rewrite => bool2int(RewriteEnabled),
+        auto_subscribe => bool2int(AutoSubEnabled),
+        delayed => bool2int(emqx:get_config([delayed, enable], false))
```
Comment (zmstone) [inline apps/emqx_modules/src/emqx_telemetry.erl:458]: emqx_delayed:is_enabled/0

File: apps/emqx_modules/src/emqx_delayed.erl [hunk 5]
```diff
@@ -147,6 +149,9 @@ enable() ->
 disable() ->
     gen_server:call(?SERVER, disable).
 
+is_enabled() ->
+    gen_server:call(?SERVER, is_enabled).
```
Comment (zmstone) [inline apps/emqx_modules/src/emqx_delayed.erl:153]: this call may fail with `timeout` exception when process is busy.
or `noproc` exception if telemetry is started earlier than `emqx_delayed`

Comment (thalesmg) [reply apps/emqx_modules/src/emqx_delayed.erl:153]: You're right.  And retainer has the same problem.  I think I'll change both to use an application env flag value to indicate that.  Thanks for pointing it out!  :beers:

**Comment by zmstone** [comment]
maybe we can move the is_enabled api to the app module
my intention was to keep the config gets close (i.e. not scattered)

**Comment by zmstone** [comment]
> maybe we can move the is_enabled api to the app module my intention was to keep the config gets close (i.e. not scattered)

for users who enable/disable on-the-fly, we can let them go.

**Comment by thalesmg** [comment]
After our discussion, I've changed the implementation to just check the configs at application start time and store the results in the application env for all future invocations.

File: apps/emqx_modules/src/emqx_modules_app.erl [hunk 6]
```diff
@@ -33,7 +33,21 @@ stop(_State) ->
     ok.
 
 maybe_enable_modules() ->
-    emqx_conf:get([delayed, enable], true) andalso emqx_delayed:enable(),
+    DelayedEnabled = emqx_conf:get([delayed, enable], true),
+    RewriteEnabled = length(emqx_conf:get([rewrite], [])) > 0,
+    RetainerEnabled = emqx_conf:get([retainer, enable], false),
+    AutoSubscribeEnabled = length(emqx_conf:get([auto_subscribe, topics], [])) > 0,
+    application:set_env(
```
Comment (zmstone) [inline apps/emqx_modules/src/emqx_modules_app.erl:40]: Set env for this app and expose an api for telemetry to access

File: apps/emqx_modules/test/emqx_telemetry_SUITE.erl [hunk 7]
```diff
@@ -150,12 +205,70 @@ t_send_after_enable(_) ->
...
+    emqx_metrics:set('messages.sent', 10_000_000_000),
+    emqx_metrics:set('messages.received', 20_000_000_000),
+    emqx_stats:setstat('topics.count', 30_000),
+    {MQTTRTInsights2, _State2} = emqx_telemetry:mqtt_runtime_insights(State1),
+    assert_approximate(MQTTRTInsights2, messages_sent_rate, "16.53"),
+    assert_approximate(MQTTRTInsights2, messages_received_rate, "33.07"),
+    ?assertEqual(30_000, maps:get(num_topics, MQTTRTInsights2)),
+    ok.
+
+assert_approximate(Map, Key, Expected) ->
```
Comment (ieQu1) [inline apps/emqx_modules/test/emqx_telemetry_SUITE.erl:268]: https://kafka4beam.github.io/snabbkaffe-docs/1.0.0-1-ga1bcd97/offline_analysis.html#give_or_take ?

File: apps/emqx_modules/test/emqx_telemetry_SUITE.erl [hunk 8]
```diff
@@ -66,17 +66,35 @@ init_per_testcase(t_get_telemetry, Config) ->
...
+    [{old_values, OldValues} | Config];
 init_per_testcase(_Testcase, Config) ->
     TestPID = self(),
     ok = meck:new(httpc, [non_strict, passthrough, no_history, no_link]),
-    ok = meck:expect(httpc, request, fun(Method, URL, Headers, Body) ->
+    ok = meck:expect(httpc, request, fun(
+        Method, {URL, Headers, _ContentType, Body}, _HTTPOpts, _Opts
+    ) ->
         TestPID ! {request, Method, URL, Headers, Body}
     end),
```
Comment (ieQu1) [inline apps/emqx_modules/test/emqx_telemetry_SUITE.erl:89]: Is it automatic formatting?

**Review by ieQu1 — DISMISSED** [review]

File: apps/emqx_modules/test/emqx_telemetry_SUITE.erl (hunk 8 above)
Comment (thalesmg) [reply apps/emqx_modules/test/emqx_telemetry_SUITE.erl:89]: Yes, I ran `scripts/format_app.py -a apps/emqx_modules -f` and it gave me this.

Comment (ieQu1) [reply apps/emqx_modules/test/emqx_telemetry_SUITE.erl:89]: :see_no_evil:

**Review by zmstone — APPROVED** [review]

File: apps/emqx_modules/src/emqx_modules_app.erl (hunk 6 above)
Comment (zmstone) [reply apps/emqx_modules/src/emqx_modules_app.erl:40]: nit: call the set api

Comment (thalesmg) [reply apps/emqx_modules/src/emqx_modules_app.erl:40]: D'oh, changed other places to use it and forgot this one... :see_no_evil:

Comment (thalesmg) [reply apps/emqx_modules/src/emqx_modules_app.erl:40]: I'll fix it in a following PR.

**Pull request merged**